src/sql/workbench/contrib/notebook/browser/outputs/gridOutput.component.ts

```diff
@@ -114,8 +121,23 @@ export class GridOutputComponent extends AngularDisposable implements IMimeCompo
 					this.updateResult(e.resultSet, e.rows);
 				}
 			}));
+			if (this._cellModel.executionState === CellExecutionState.Running || !this._incrementalGridRenderingEnabled) {
+				await this.renderGrid();
+			} else {
+				this.setLoading(true);
+				setTimeout(async () => {
```

Comment:
What's the setTimeout for? Could you not just do 

`this.renderGrid().then( () => this.setLoading(false));` and then make sure to add a comment about purposely not awaiting this?

(you could also just do setLoading to false in the renderGrid call - that should be fine to do even if incremental loading isn't enabled)

Reply:
setTimeout will add the renderGrid() call to a queue that gets called after all current functions get executed - so this lets the rest of the notebook render first.
When I do
``` 
this.renderGrid().then( () => this.setLoading(false));
```
nothing is rendered and ADS is frozen until all things are added to the DOM.

Reply:
That seems odd - do you know why that happens? Is it getting stuck on one of these calls?

```
		if (!this._table) {
			let source = <IDataResource><any>this._bundleOptions.data[this.mimeType];
			reorderGridData(source);
			let state = new GridTableState(0, 0);
			this._table = this.instantiationService.createInstance(DataResourceTable, source, this.cellModel, this.cellOutput, state);
			let outputElement = <HTMLElement>this.output.nativeElement;
			outputElement.appendChild(this._table.element);
			this._register(attachTableStyler(this._table, this.themeService));
```

Would be interesting to know where this is getting stuck - because once it hits the await then it should be letting other tasks be run while it loads stuff.

Reply:
Anyways you don't have to spend a bunch of time on this right now - it's not super important. setTimeout is fine - we do something similar in other places where we want to force a context switch. Just make sure to add comments explaining that for future reference.

Reply:
According to this: https://javascript.info/event-loop,
"Rendering never happens while the engine executes a task. It doesn’t matter if the task takes a long time. Changes to the DOM are painted only after the task is complete."

Without the setTimeout, the entire notebook rendering is considered one task so nothing gets rendered until the very end. Whereas using setTimeout will break up the grid rendering into separate tasks so each time one grid is rendered, we can see the updated DOM.

Reply:
Ahhh, ok so I didn't know about `macrotasks` vs `microtasks`

https://javascript.info/event-loop#macrotasks-and-microtasks

That's the key here - promise execution IS a separate task but it's a microtask, so it'll still be completed before moving on to rendering. Where as setTimeout creates a new macrotask and so will only happen after it does another render call.

@alanrenmsft FYI - you might find this interesting